### PR TITLE
Implement if-else in wasm_to_x86 backend

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**
Related #1222

### CODE REVIEW

File: src/libasr/codegen/wasm_to_x86.cpp
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/wasm_to_x86.cpp:84]: I have a doubt:
Does `wasm` always have an else block (even though else is empty) in the if statement?
image
@Shaikh-Ubaid 
If so how Assert is handled in asr_to_wasm?

Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/wasm_to_x86.cpp:84]: I asked this because, the else block is not empty in the `visit_assert()`, it has `0x41`

Comment (Shaikh-Ubaid) [reply src/libasr/codegen/wasm_to_x86.cpp:84]: >Does wasm always have an else block (even though else is empty) in the if statement?

Currently, if there is no `else` block, then in the `wat` generated, we would just have `if` `<some instructions>` `end`.

Comment (Shaikh-Ubaid) [reply src/libasr/codegen/wasm_to_x86.cpp:84]: >I asked this because, the else block is not empty in the visit_assert(), it has 0x41

Do you mean `visit_Assert()` in `ASR->WASM` backend?

Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/wasm_to_x86.cpp:84]: yes

File: src/libasr/codegen/wasm_to_x86.cpp
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/wasm_to_x86.cpp:91]: But this logic, detects the else block, even for the following syntax!!
```
if True:
	pass
```

Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/wasm_to_x86.cpp:91]: That is the reason, I check the next byte to be 0x0B (which checks the end of the if stmt).

Comment (Shaikh-Ubaid) [reply src/libasr/codegen/wasm_to_x86.cpp:91]: At present, if there is no `else` block, then `visit_Else()` should not be visited. I guess, we might be missing something.

**Comment by Shaikh-Ubaid** [comment]
I looked into the `PR`. It seems, it would be better if `wasm` backend adds an `else` block always (it would be empty when it is not needed).

If an `else` is always present, then there would not be any need of checking if any `else` is present (since, it would be always present (although empty when not needed)).

**Comment by Shaikh-Ubaid** [comment]
I will rebase this on top of https://github.com/lcompilers/lpython/pull/1262 (so that we can print some value inside the `if-else` ladder), make the necessary updates to the `wasm` backend and push to this `PR`.

**Comment by Thirumalai-Shaktivel** [comment]
This looks great! Thanks for updating.

**Comment by Shaikh-Ubaid** [comment]
>I will rebase this on top of https://github.com/lcompilers/lpython/pull/1262

I rebased on top of latest `main`. Rebasing on top of https://github.com/lcompilers/lpython/pull/1262 is including many commits of that `PR` (it also makes this PR dependent on it).

**Review by Shaikh-Ubaid — APPROVED** [review]

**Comment by Thirumalai-Shaktivel** [comment]
Let's implement visit_Assert() in this PR itself.
Assertion in wasm_x86, behaves differently!
```py
def Main0():
    assert 1 == 1

Main0()
```
Output:
```console
$ lpython examples/expr2.py --backend wasm_x86 && ./a.out
AssertionError
```
```py
def Main0():
    assert 0 == 1

Main0()
```
```console
$ lpython examples/expr2.py --backend wasm_x86 && ./a.out
```

**Comment by Thirumalai-Shaktivel** [comment]
@Shaikh-Ubaid What is EmptyBlockType? How is Assert handling in `WASM` (it's seems wasm generates correct output)?

File: src/libasr/codegen/asr_to_wasm.cpp
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_wasm.cpp:2113]: The previous version was more correct, right? The previous version would say -> If `Assert` condition failed, raise an error (which is what `assert` does, I think).

File: src/libasr/codegen/asr_to_wasm.cpp
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_wasm.cpp:2122]: The `else` portion was empty in the previous version.

**Review by Shaikh-Ubaid — CHANGES_REQUESTED** [review]
The right way to support `Assert` would instead be to implement the `visit_I32Eqz` in the `WASM->X86` backend.

```cpp
void visit_I32Eqz() {
        m_a.asm_pop_r32(X86Reg::eax);
        m_a.asm_mov_r32_imm32(X86Reg::ebx, 0U);
        m_a.asm_cmp_r32_r32(X86Reg::eax, X86Reg::ebx);
        m_a.asm_push_r32(X86Reg::eax);
    }
```

**Comment by certik** [comment]
Thanks @Thirumalai-Shaktivel for working on this, very helpful. Thanks @Shaikh-Ubaid for the review.

File: src/libasr/codegen/asr_to_wasm.cpp
Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/asr_to_wasm.cpp:2113]: The main motive for these changes was that `visit_I32Eqz()` in the `wasm_to_x86` is not working correctly.
Here, the changes implies:
```py
If True:
	// Do nothing
else:
	print("AssertionError")
```
which is similar to `assert False`

**Comment by Thirumalai-Shaktivel** [comment]
> The right way to support `Assert` would instead be to implement the `visit_I32Eqz` in the `WASM->X86` backend.
> 
> ```c++
> void visit_I32Eqz() {
>         m_a.asm_pop_r32(X86Reg::eax);
>         m_a.asm_mov_r32_imm32(X86Reg::ebx, 0U);
>         m_a.asm_cmp_r32_r32(X86Reg::eax, X86Reg::ebx);
>         m_a.asm_push_r32(X86Reg::eax);
>     }
> ```

This seems not working in wasm_x86 backend.

File: src/libasr/codegen/asr_to_wasm.cpp
Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/asr_to_wasm.cpp:2113]: The same approach as we do in the LLVM backend: https://github.com/lcompilers/lpython/blob/d4a6d5360ca520d9a25d658e6660cb39a2eda4e2/src/libasr/codegen/asr_to_llvm.cpp#L4845-L4862

**Comment by Thirumalai-Shaktivel** [comment]
What does I32Eqz do? isn't it a NOT operator?

**Comment by Shaikh-Ubaid** [comment]
>What does I32Eqz do? isn't it a [NOT](https://github.com/sunfishcode/wasm-reference-manual/blob/master/WebAssembly.md#integer-equal-to-zero) operator?

It checks if the stack top is equal to zero. Yes, it works similar to a not operator.

>The same approach as we do in the LLVM backend: https://github.com/lcompilers/lpython/blob/d4a6d5360ca520d9a25d658e6660cb39a2eda4e2/src/libasr/codegen/asr_to_llvm.cpp#L4845-L4862

If it is similar to the `LLVM` backend, we can follow it.



```
void visit_I32Eqz() {
        m_a.asm_pop_r32(X86Reg::eax);
        m_a.asm_mov_r32_imm32(X86Reg::ebx, 0U);
        m_a.asm_cmp_r32_r32(X86Reg::eax, X86Reg::ebx);
        m_a.asm_push_r32(X86Reg::eax);
}.
```
>This seems not working in the wasm_x86 backend

It should (hopefully) work. I tested it locally on my system before sharing here. If we are following the `LLVM` approach, then we might not need it (I guess) and we can add it later when needed.

**Comment by Thirumalai-Shaktivel** [comment]
> It should (hopefully) work. I tested it locally on my system before sharing here.

Were you able to get "AssertionError" for `assert False`?
This is weird, it didn't work for me!

**Comment by Thirumalai-Shaktivel** [comment]
```py
def Test_assert_01():
    if True:
        assert True

    assert 1 != 1, "One is equal to one."

def Verify():
    Test_assert_01()

Verify()
```
```console
$ lpython examples/expr2.py --backend wasm_x86 && ./a.out
Call to imported function with index 4 not yet supported
Call to imported function with index 4 not yet supported
```

**Comment by Thirumalai-Shaktivel** [comment]
Sorry my bad, print_str doesn't work yet in this PR! I will rebase that PR and report back

**Comment by Shaikh-Ubaid** [comment]
>Sorry my bad, print_str doesn't work yet in this PR! I will rebase that PR and report back.

Let's wait for the `PRs` to be merged. It seems many `PRs` are interdependent. Let's work on `Assertion` in another `PR` once all the necessary `PRs` are merged.

**Comment by Thirumalai-Shaktivel** [comment]
Yup! Now, I feel the same.
I will another PR for `Assert` once the `print_str` is merged

**Review by Shaikh-Ubaid — APPROVED** [review]

**Comment by Thirumalai-Shaktivel** [comment]
Thank you for your approval and guidance!

File: src/libasr/codegen/wasm_to_x86.cpp
Comment (certik) [inline src/libasr/codegen/wasm_to_x86.cpp:19]: Later:
```suggestion
    uint32_t unique_id;
```

File: src/libasr/codegen/wasm_to_x86.cpp
Comment (certik) [inline src/libasr/codegen/wasm_to_x86.cpp:75]:
```suggestion
        unique_id++;
        uint32_t label = unique_id;
```

File: src/libasr/codegen/wasm_to_x86.cpp
Comment (certik) [inline src/libasr/codegen/wasm_to_x86.cpp:80]: For later (but probably quite soon): we should not be using strings for labels, but just unique integers. The unique integer can just be a counter in this visitor, just an integer, see above.

Comment (certik) [reply src/libasr/codegen/wasm_to_x86.cpp:80]: (we might need to have a stack of unique ids...)

**Review by certik — APPROVED** [review]
Looks good to me, thanks!

**Pull request merged**
